site/content/entry/citra-progress-report-2017-september.md

```diff
@@ -0,0 +1,360 @@
+is mapped to memory or IO. Unfortunately, this is extremely slow, and we can't
+afford to have extremely slow address translation when games can access memory
+upwards of a few hundred thousand times per second.
+
+With this, [MerryMage](https://github.com/MerryMage) has changed this behaviour so
+that rather than replacing a read/write with a function, it instead translates the
+address using a page table, and then tries to access that address directly. On the
+page table, all addresses that map to memory simply have a memory address written down.
+But on addresses that map to IO, it has address 0 written down. Trying to read or
+write to memory address 0 on x86 is illegal for every process except the
+operating system... and Citra tries to do it anyways!
+
+When an invalid memory address (or a memory address that that process doesn't have
+permission to access) is read from or written to, x86 CPUs throw a page fault exception.
+Citra takes advantage of this behaviour by also registering an exception handler
+for page faults. If a page fault is thrown, Citra knows the game tried to access IO,
+and thus recompiles the memory read/write to a direct call to Citra's IO functions.
+This makes the usual case (memory access) extremely fast, and the less usual case
+slow, but only the first time it happens. Subsequent IO accesses use the recompiled
+functions which are faster.
+
+This technique is called fastmem, and is not new at all. In fact, Dolphin uses
+it extensively in its JIT recompiler to speed up memory access as well. And thanks
+to [MerryMage](https://github.com/MerryMage)'s hard work, this same technique is
+now used extensively by Citra.
+
+## [Give each process its own page table](https://github.com/citra-emu/citra/pull/2842) by [Subv](https://github.com/Subv)
+
+In order to support running multiple processes at the same time, like your computer,
+Citra implements virtual memory, in which each process has its own page table.
+The page table represents a translation from the process' virtual addresses, to
+the 3DS' physical (or "real") addresses.
+
+Before this, because Citra did not support multiple page tables, it also didn't
+support running multiple processes at once, such as a game and the software keyboard
+applet. Now, thanks to [Subv](https://github.com/Subv), Citra has an important
+building block in place.
+
+## [Add support for loading application updates](https://github.com/citra-emu/citra/pull/2927) by [shinyquagsire23](https://github.com/shinyquagsire23)
+
+Nintendo 3DS titles are contained within `*.app` files on the SD card or on the
+game cartridge, in the [NCCH container format](https://www.3dbrew.org/wiki/NCCH).
+This format is further divided into two formats, CXI and CFA, which stand for
+__C__TR e__X__ecutable __I__mage and __C__TR __F__ile __A__rchive, respectively.
+CXIs contain executable code, whereas CFAs cannot. CFAs usually accompany a CXI
```

Comment:
this isn't true, right? CFA's can contain executable code, they are just not directly bootable, right?

Reply:
CFAs don't have the executable flag, though they *can* contain code. DLP uses this behaviour to store the child CIA. Was trying to KISS for readers.